plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ForwardingConnection.java

```diff
@@ -0,0 +1,478 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.plugin.jdbc;
+
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.CallableStatement;
+import java.sql.Clob;
+import java.sql.Connection;
+import java.sql.DatabaseMetaData;
+import java.sql.NClob;
+import java.sql.PreparedStatement;
+import java.sql.SQLClientInfoException;
+import java.sql.SQLException;
+import java.sql.SQLWarning;
+import java.sql.SQLXML;
+import java.sql.Savepoint;
+import java.sql.ShardingKey;
+import java.sql.Statement;
+import java.sql.Struct;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.Executor;
+
+public abstract class ForwardingConnection
```

Comment:
Yes, it is, but this looks like a simple delegation everywhere (apart from the exception irregularities). Do you expect that one you have some more experience with how this works you'll have to override some of the methods with custom implementation?

Reply:
It is not that simple. See `setClientInfo`. Possibly in future one may want to change `unwrap` too.

I prefer forwarding as it is has better stack trace and it is easier to deal with exceptions. In most case I just pass the exception further.

Also I prefer to avoid reflection if it is possible. This is one time effort and it is clear what is the mapping. If I would saw reflection here then I would ask myself if there are no issues with it. Not the case here.

Reply:
My reaction would be exactly the opposite - I see all those forwarding calls and I wonder if there are any mistakes somewhere in there. On the other hand, I worked with dynamic proxies quite a lot before, so I'm familiar with all their quirks. But I see your point - no further objection from my side.

Reply:
@ksobolew see eg `com.google.common.collect.ForwardingMap`.

it's as dumb forwarding as possible, given checked exceptions, which is inherent consequence of the API we're forwarding